### ISSUE TITLE
FreeBSD: fs_event_watch_delete_dir segfaults

### ISSUE THREAD

**Issue by rhelmot** [libuv/libuv#4606]
Possibly related to https://github.com/libuv/libuv/issues/4255

```
not ok 61 - fs_event_watch_delete_dir                                                                                  
# exit code 139                                                                                                        
# Output from process `fs_event_watch_delete_dir`: (no output) 
```

gdbing indicates that filename is a null pointer [here](https://github.com/libuv/libuv/blob/v1.49.2/test/test-fs-event.c#L156).

* **Version**: libuv 1.49.2
* **Platform**: FreeBSD 14.0

**Comment by bnoordhuis** [comment]
cc @devnexen - likely related to 004dfd2d4b2bc3c3f2ddda7ac72c9dd579ca4194?

**Comment by devnexen** [comment]
WIll have a look soon-ish.

**Comment by rhelmot** [comment]
with that diff, the error changes to the following:

```
# exit code 134                              
# Output from process `fs_event_watch_dir`:
# Assertion failed in test/test-fs-event.c on line 233: `filename == file_prefix` (watch_di == fsevent-) 
```

**Comment by juanarbol** [comment]
can Confirm. I will take a look to this one.

**Comment by juanarbol** [comment]
@libuv/collaborators, this seems to be a kernel [bug](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=197695).

Refs: https://stackoverflow.com/a/28778077, https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=198570

**Comment by juanarbol** [comment]
As #4649 landed, is it safe to close this?

Cc @cjihrig

**Comment by cjihrig** [comment]
> is it safe to close this?

I'm not sure. It seems like the discussion was previously headed towards a workaround for the bug. @bnoordhuis what do you think?

**Comment by bnoordhuis** [comment]
Downstream users won't expect sneaky NULLs. It'd be best to fix this on our side.

**Comment by juanarbol** [comment]
I'll investigate.

@bnoordhuis, If you have a preferred workaround, let me know

**Comment by bnoordhuis** [comment]
https://github.com/libuv/libuv/issues/4606#issuecomment-2486459135 is what we do on other platforms, IIRC, so I'd just copy that for freebsd.

**Comment by juanarbol** [comment]
> with that diff, the error changes to the following:
> 
> 
> 
> ```
> 
> # exit code 134                              
> 
> # Output from process `fs_event_watch_dir`:
> 
> # Assertion failed in test/test-fs-event.c on line 233: `filename == file_prefix` (watch_di == fsevent-) 
> 
> ```

That will cause another crash.

**Comment by ptlomholt** [comment]
The documentation at https://docs.libuv.org/en/v1.x/fs_event.html#c.uv_fs_event_cb suggests that it _can_ be NULL: 'the filename parameter will be a relative path to a file contained in the directory, or NULL if the file name cannot be determined' so maybe it is ok...

**Comment by juanarbol** [comment]
> [#4606 (comment)](https://github.com/libuv/libuv/issues/4606#issuecomment-2486459135) is what we do on other platforms, IIRC, so I'd just copy that for freebsd.

@bnoordhuis I don't think that is gonna fix anything, the test seems to be conveniently ignoring when the -whole- file path is not determined (like the FreeBSD case)[1], if we do that, we will get the path of the directory, but no the the -whole- file route.

I don't think there is reliable way to get the path of an fd in FreeBSD (without making extra requirements to the user's environment, like mounting `/proc` or something)

[1]: https://github.com/libuv/libuv/blob/e8969bff6cfebf950b2a52eabff53e53ff02ab0e/test/test-fs-event.c#L239
